### ISSUE TITLE
some bugs need to fix in cli params subcommand

### ISSUE THREAD

**Issue by atenjin** [paritytech/substrate#1342]
I think there are some bugs in cli subcommand
1. in `core/cli/lib.rs`, func `execute_default` (L411)
I think it seems that for executing subcommand like `export-blocks`, `revert` or else, it not pass the sub_matches params, instand, it passes the main matches. In my test, if pass the main matches, I couldn't get the params in sub_matches. 
```rust
	} else if let Some(sub_matches) = matches.subcommand_matches("export-blocks") {
		export_blocks::<F, _>(
			get_db_path_for_subcommand(matches, sub_matches)?,
			matches, // notice this, not use sub_matches
			spec,
			exit.into_exit()
		)?;
		return Ok(Action::ExecutedInternally);
	} else if let Some(sub_matches) = matches.subcommand_matches("import-blocks") {
		import_blocks::<F, _>(
			get_db_path_for_subcommand(matches, sub_matches)?,
			matches, // notice this, not use sub_matches
			spec,
			exit.into_exit()
		)?;
		return Ok(Action::ExecutedInternally);
	} else if let Some(sub_matches) = matches.subcommand_matches("revert") {
		revert_chain::<F>(
			get_db_path_for_subcommand(matches, sub_matches)?,
			matches,  // notice this, not use sub_matches
			spec
		)?;
		return Ok(Action::ExecutedInternally);
	} else if let Some(sub_matches) = matches.subcommand_matches("purge-chain") {
		purge_chain::<F>(get_db_path_for_subcommand(matches, sub_matches)?)?; // this is current
		return Ok(Action::ExecutedInternally);
	}
```
For example revert, I what to revert 1 block, so I do this exec:
```shell
cargo run -- --dev -d="somepath" revert 1
```
I do debug in program, in L579
```rust
fn revert_chain<F>(
	db_path: PathBuf,
	matches: &clap::ArgMatches,
	spec: ChainSpec<FactoryGenesis<F>>
) -> error::Result<()>
	where F: ServiceFactory,
{
	let mut config = service::Configuration::default_with_spec(spec);
	config.database_path = db_path.to_string_lossy().into();

	let blocks = match matches.value_of("num") {  // I can't get num in matches, due to this matches is main matches
		Some(v) => v.parse().map_err(|_| "Invalid block count specified")?,
		None => 256,
	};

	Ok(service::chain_ops::revert_chain::<F>(config, As::sa(blocks))?)
}
```
so that it return 256 for `blocks`
I submit pr #1341 for this, but I'm not sure the subcommand `export-blocks`, `import-blocks` and `purge-chain` also need this.

2. the subcommand do `get_db_path_for_subcommand` to get a new path for db, but in node cli, I find that the path would generate by `load_spec` for chainid
in `node/cli/lib.rs` L125
```rust
	let (spec, mut config) = cli::parse_matches::<service::Factory, _>(
		load_spec, version, "substrate-node", &matches
	)?;
```
the `load_spec` would allow user to specify config by params, for example the `development_config` would specify the chainid is "dev", and the db path would generate to "dbpath/dev".
but in subcommand `get_db_path_for_subcommand`
```rust
fn get_db_path_for_subcommand(
	main_cmd: &clap::ArgMatches,
	sub_cmd: &clap::ArgMatches
) -> error::Result<PathBuf> {
        // ...
	check_contradicting_chain_dev_flags(main_cmd, sub_cmd)?;
	check_contradicting_chain_dev_flags(sub_cmd, main_cmd)?;
        // just get chain key directly, not from `load_spec`
	let spec_id = if sub_cmd.is_present("chain") || sub_cmd.is_present("dev") {
		get_chain_key(sub_cmd)
	} else {
		get_chain_key(main_cmd)
	};
```
so that if I specify another chainid in config, it would not get the chainid currently in `get_db_path_for_subcommand`
I think the `load_spec` would pass from `node` to the `core`, and let the `get_db_path_for_subcommand` to this `load_spec` to get current chainid

**Comment by JKingdom** [comment]
maybe this issue should be opened, my pr #1341 just fix "revert" subcommand, I'm not sure other subcommand also need this.
and for `get_db_path_for_subcommand`, I'm sure it's a bug, but not fix in pr #1341

**Comment by bkchr** [comment]
I'm reworking all of it anyway currently.
